src/os/filestore/JournalingObjectStore.cc

```diff
@@ -183,70 +198,103 @@ void JournalingObjectStore::ApplyManager::add_waiter(uint64_t op, Context *c)
+      if (committing_seq + 1 == it->first) {
+        if (it->second) {
+          ret = true;
+          committing_seq++;
+        } else {
+          break;
+        }
+      } else {
+        break;
       }
+    }
+    if (ret) {
+      applied_seq_map.erase(applied_seq_map.begin(), it);
+      dout(10) << "commit_start:fast_sync committing_seq = " << committing_seq << dendl;
+    }
 
-      _committing_seq = committing_seq = max_applied_seq;
+  } else {
 
-      dout(10) << "commit_start committing " << committing_seq
+    uint64_t _committing_seq = 0;
+    {
+      Mutex::Locker l(apply_lock);
+      dout(10) << "commit_start max_applied_seq " << max_applied_seq
+    	     << ", open_ops " << open_ops
+	     << dendl;
+      blocked = true;
+      while (open_ops > 0) {
+        dout(10) << "commit_start waiting for " << open_ops << " open ops to drain" << dendl;
+        blocked_cond.Wait(apply_lock);
+      }
+      assert(open_ops == 0);
+      dout(10) << "commit_start blocked, all open_ops have completed" << dendl;
+      {
+        Mutex::Locker l(com_lock);
+        if (max_applied_seq == committed_seq) {
+	  dout(10) << "commit_start nothing to do" << dendl;
+	  blocked = false;
+	  assert(commit_waiters.empty());
+	  goto out;
+        }
+
+        _committing_seq = committing_seq = max_applied_seq;
+
+        dout(10) << "commit_start committing " << committing_seq
 	       << ", still blocked" << dendl;
+      }
     }
-  }
-  ret = true;
+    ret = true;
 
- out:
-  if (journal)
-    journal->commit_start(_committing_seq);  // tell the journal too
+   out:
+    if (journal)
+      journal->commit_start(_committing_seq);  // tell the journal too
+  }
   return ret;
 }
 
-void JournalingObjectStore::ApplyManager::commit_started()
+void JournalingObjectStore::ApplyManager::commit_started(bool fast_sync)
 {
-  Mutex::Locker l(apply_lock);
-  // allow new ops. (underlying fs should now be committing all prior ops)
-  dout(10) << "commit_started committing " << committing_seq << ", unblocking" << dendl;
-  blocked = false;
-  blocked_cond.Signal();
+  if (!fast_sync) {
+    Mutex::Locker l(apply_lock);
+    // allow new ops. (underlying fs should now be committing all prior ops)
+    dout(10) << "commit_started committing " << committing_seq << ", unblocking" << dendl;
+    blocked = false;
+    blocked_cond.Signal();
+  }
 }
 
-void JournalingObjectStore::ApplyManager::commit_finish()
+void JournalingObjectStore::ApplyManager::commit_finish(bool fast_sync)
 {
-  Mutex::Locker l(com_lock);
-  dout(10) << "commit_finish thru " << committing_seq << dendl;
-
-  if (journal)
-    journal->committed_thru(committing_seq);
-
-  committed_seq = committing_seq;
-
-  map<version_t, vector<Context*> >::iterator p = commit_waiters.begin();
-  while (p != commit_waiters.end() &&
-    p->first <= committing_seq) {
-    finisher.queue(p->second);
-    commit_waiters.erase(p++);
+  if (fast_sync) {
+    if (journal)
```

Comment:
Why don't we need to take the com_lock here?  This branch also seems to be identical to the beginning of the next one except for the lock.